python_modules/dagster-graphql/dagster_graphql/implementation/fetch_assets.py

```diff
@@ -400,20 +401,39 @@ def get_asset_observations(
     return [event_record.event_log_entry for event_record in event_records]
 
 
-def get_assets_for_run_id(graphene_info: "ResolveInfo", run_id: str) -> Sequence["GrapheneAsset"]:
+def get_assets_for_run(graphene_info: "ResolveInfo", run: DagsterRun) -> Sequence["GrapheneAsset"]:
     from dagster_graphql.schema.pipelines.pipeline import GrapheneAsset
 
-    check.str_param(run_id, "run_id")
+    run_id = run.run_id
+
+    # Fetch observations and materialization events from run
+    event_log_records = [
+        *graphene_info.context.instance.get_records_for_run(
+            run_id,
+            of_type=DagsterEventType.ASSET_MATERIALIZATION,
+        ).records,
+        *graphene_info.context.instance.get_records_for_run(
+            run_id,
+            of_type=DagsterEventType.ASSET_OBSERVATION,
+        ).records,
+    ]
+
+    asset_keys = {
+        record.event_log_entry.dagster_event.asset_key
+        for record in event_log_records
+        if record.event_log_entry.dagster_event
+    }
 
-    records = graphene_info.context.instance.all_logs(run_id, of_type=ASSET_EVENTS)
-    asset_keys = set(
-        [
-            record.get_dagster_event().asset_key
-            for record in records
-            if record.is_dagster_event and record.get_dagster_event().asset_key
-        ]
-    )
-    return [GrapheneAsset(key=asset_key) for asset_key in asset_keys]
+    # Fetch planned asset keys from execution plan snapshot
+    if run.execution_plan_snapshot_id:
+        execution_plan_snapshot = check.not_none(
+            graphene_info.context.instance.get_execution_plan_snapshot(
+                run.execution_plan_snapshot_id
+            )
+        )
+        asset_keys.update(execution_plan_snapshot.asset_selection)
```

Comment:
this also came up in the PR that added ExecutionPlanSnapshot.asset_selection in the first place (for a similar purpose: https://github.com/dagster-io/dagster/pull/23179

The problem with matching the logic that emits the events is that the logic that emits the event also needs to introspect the outputs to do various other things with them, so i couldn't find a great way to pull it into a shared helper function. It does match though!

Reply:
It's not exactly the same as step_keys_to_execute because not every step has an asset_key, but it is the subset of step_keys_to_execute with an asset key

Reply:
gotcha, thanks for explaining